File: packages/naver-map/ios/NaverMapRegistry.swift
//
//  NaverMapRegistry.swift
//  granite-naver-map
//
//  Registry singleton for NaverMap providers
//

import Foundation

@objc public class NaverMapRegistry: NSObject {
    @objc public static let shared = NaverMapRegistry()

    @objc public private(set) var provider: NaverMapProvidable?

    private override init() {
        super.init()
    }

    /// Register a custom provider. Call this at app startup before using NaverMap.
    @objc public func register(provider: NaverMapProvidable) {
        self.provider = provider
    }

    /// Get the current provider, falling back to built-in provider if none registered
    @objc public func getProvider() -> NaverMapProvidable {
        if let provider = provider {
            return provider
        }

        // Auto-register built-in provider on first use
        let builtIn = BuiltInNaverMapProvider()
        self.provider = builtIn
        return builtIn
    }

    /// Check if a custom provider has been registered
    @objc public var hasCustomProvider: Bool {
        return provider != nil
    }

    /// Reset the provider (useful for testing)
    @objc public func reset() {
        provider = nil
    }
}

File: packages/naver-map/android/src/main/java/run/granite/navermap/NaverMapRegistry.kt
package run.granite.navermap

import android.content.Context

/**
 * Registry singleton for NaverMap providers
 *
 * Brownfield apps can register their own provider implementation at app startup.
 * If no provider is registered, the built-in NMapsMap provider will be used.
 *
 * Usage:
 * ```kotlin
 * // In your Application.onCreate() or before using NaverMap:
 * NaverMapRegistry.register(MyCustomNaverMapProvider())
 * ```
 */
object NaverMapRegistry {
    private var _provider: NaverMapProvider? = null

    /**
     * The currently registered provider
     */
    val provider: NaverMapProvider?
        get() = _provider

    /**
     * Register a custom provider. Call this at app startup before using NaverMap.
     */
    fun register(provider: NaverMapProvider) {
        _provider = provider
    }

    /**
     * Get the current provider, creating the built-in provider if none registered
     */
    fun getProvider(context: Context): NaverMapProvider {
        return _provider ?: BuiltInNaverMapProvider(context).also {
            _provider = it
        }
    }

    /**
     * Check if a custom provider has been registered
     */
    val hasCustomProvider: Boolean
        get() = _provider != null

    /**
     * Reset the provider (useful for testing)
     */
    fun reset() {
        _provider = null
    }
}
